main.tf

```diff
@@ -1,3 +1,132 @@
-*********************************************************************/
+###########################################################
+# This file creates an event notificaiton resource instance
+###########################################################
+locals {
+  # tflint-ignore: terraform_unused_declarations
+  validate_kms_plan = var.kms_encryption_enabled && var.plan != "standard" ? tobool("kms encryption is only supported for standard plan") : true
+  # tflint-ignore: terraform_unused_declarations
+  validate_kms_values = !var.kms_encryption_enabled && (var.kms_key_crn != null || var.root_key_id != null) ? tobool("When passing values for var.kms_key_crn or/and var.root_key_id, you must set var.kms_encryption_enabled to true. Otherwise unset them to use default encryption") : true
+  # tflint-ignore: terraform_unused_declarations
+  validate_kms_vars = var.kms_encryption_enabled && (var.kms_key_crn == null || var.root_key_id == null) ? tobool("When setting var.kms_encryption_enabled to true, a value must be passed for var.kms_key_crn and var.root_key_id") : true
+  # tflint-ignore: terraform_unused_declarations
+  validate_auth_policy = var.kms_encryption_enabled && var.skip_iam_authorization_policy == false && var.existing_kms_instance_guid == null ? tobool("When var.skip_iam_authorization_policy is set to false, and var.kms_encryption_enabled to true, a value must be passed for var.existing_kms_instance_guid in order to create the auth policy.") : true
+
+  # Determine what KMS service is being used for encryption
+  kms_service = var.kms_key_crn != null ? (
+    can(regex(".*kms.*", var.kms_key_crn)) ? "kms" : (
+      can(regex(".*hs-crypto.*", var.kms_key_crn)) ? "hs-crypto" : null
+    )
+  ) : null
+}
+
+resource "ibm_resource_instance" "en_instance" {
+  plan              = var.plan
+  location          = var.region
+  name              = var.name
+  resource_group_id = var.resource_group_id
+  tags              = var.tags
+  service           = "event-notifications"
+
+  parameters = {
+    service-endpoints = var.service_endpoints
+  }
+}
+
+#############################################################################
+# Event Notification KMS integration
+#############################################################################
+
+locals {
+  en_integration_id = [
+    for integrations in data.ibm_en_integrations.en_integrations.integrations :
+    integrations.id if integrations.type == local.kms_service
+  ]
+}
+data "ibm_en_integrations" "en_integrations" {
+  instance_guid = ibm_resource_instance.en_instance.guid
+}
+
+resource "ibm_en_integration" "en_kms_integration" {
+  count          = var.kms_encryption_enabled == false || var.skip_iam_authorization_policy ? 0 : 1
+  instance_guid  = ibm_resource_instance.en_instance.guid
+  integration_id = local.en_integration_id[0]
+  type           = local.kms_service
+  metadata {
+    endpoint    = "https://${var.region}.kms.cloud.ibm.com"
```

Comment:
The KMS region could be a different region to the Even Notifications instance. So you need a new `kms_region` variable for this. 

Also this logic only supports the public endpoint. You will need to add a new variable called `kms_endpoint` with valid values of `public` or `private`, and then update the endpoint value here to 
`var.kms_endpoint == "public" ? "https://${var.kms_region}.kms.cloud.ibm.com" : "https://private.${var.kms_region}.kms.cloud.ibm.com"`

Reply:
Done.